internal/service/database/database.go

```diff
@@ -12,427 +12,548 @@ import (
-		"maintenance_window_dow", "maintenance_window_time", "properties.0", "network", "labels", "additional_disk_space_gib") {
-		req := request.ModifyManagedDatabaseRequest{UUID: d.Id()}
-		req.Plan = d.Get("plan").(string)
-		req.Title = d.Get("title").(string)
-		req.Zone = d.Get("zone").(string)
-		if d.HasChanges("maintenance_window_dow", "maintenance_window_time") {
-			req.Maintenance.DayOfWeek = d.Get("maintenance_window_dow").(string)
-			req.Maintenance.Time = d.Get("maintenance_window_time").(string)
-		}
+func setDatabaseProperties(ctx context.Context, data *databaseCommonModel, db *upcloud.ManagedDatabase, isImport bool) diag.Diagnostics {
+	var diags, d diag.Diagnostics
 
-		if d.HasChanges("additional_disk_space_gib") {
-			additionalDiskSpaceGiB := d.Get("additional_disk_space_gib").(int)
-			req.AdditionalDiskSpaceGiB = &additionalDiskSpaceGiB
-		}
+	propsInfo := properties.GetProperties(db.Type)
+	propsData := make(map[string]attr.Value)
 
-		if d.HasChange("labels") {
-			labels := utils.LabelsMapToSlice(d.Get("labels").(map[string]interface{}))
-			req.Labels = &labels
-		}
+	prevProps, err := properties.ListToValueMap(ctx, data.Properties)
+	if err != nil {
+		diags.AddError(
+			"Unable to parse managed database properties from plan",
+			utils.ErrorDiagnosticDetail(err),
+		)
+		return diags
+	}
 
-		if d.HasChange("termination_protection") {
-			terminationProtection := d.Get("termination_protection").(bool)
-			req.TerminationProtection = &terminationProtection
-		}
+	for typedKey, value := range db.Properties {
+		key := string(typedKey)
 
-		if d.HasChange("properties.0") {
-			props := buildManagedDatabasePropertiesRequestFromResourceData(d)
+		// Skip properties that are not defined in the propsInfo
+		prop, ok := propsInfo[key]
+		if !ok {
+			continue
+		}
 
-			// Always delete version if it exists; versions are updated via separate endpoint
-			delete(props, "version")
-			req.Properties = props
+		// Create-only properties are handled by plan-modifiers
+		if prop.CreateOnly {
+			continue
 		}
 
-		if d.HasChange("network") {
-			networks := networksFromResourceData(d)
-			req.Networks = &networks
+		// Skip properties that are null in the plan
+		if !isImport && prevProps[properties.SchemaKey(key)].IsNull() {
+			continue
 		}
 
-		_, err := client.ModifyManagedDatabase(ctx, &req)
+		// Ignore known differences between API and plan values
+		processedValue, err := ignorePropChange(value, prevProps[properties.SchemaKey(key)], key, prop)
 		if err != nil {
-			return diag.FromErr(err)
+			diags.AddError(
+				"Unable to process managed database property value",
+				utils.ErrorDiagnosticDetail(err),
+			)
+			return diags
 		}
 
-		tflog.Info(ctx, "managed database updated", map[string]interface{}{"uuid": d.Id(), "name": d.Get("name")})
-	}
-
-	// Modify powered state if no PostgreSQL version update is requested
-	if d.HasChange("powered") && !d.HasChange("properties.0.version") {
```

Comment:
Removed this special case as I assume the DB system would throw some errors if trying to upgrade version right after start of shut-down right after version upgrade without waiting for stable state between these actions 🤔